src/windows.rs

```diff
@@ -0,0 +1,464 @@
+            Err(e) if e == ERROR_NOT_FOUND || e == WIN32_ERROR(6) => {
+                ddcci_get_monitor_brightness(self)
+                    .map(|b| b.get_current_percentage())
+                    .map_err(|e| e.into())
+            }
+            Err(e) => Err(SysError::GettingBrightnessFailed {
+                device_name: self.device_name.clone(),
+                function_name: "IOCTL_VIDEO_QUERY_SUPPORTED_BRIGHTNESS",
+                last_error: e,
+            }
+            .into()),
+        };
+    }
+
+    async fn set(&mut self, percentage: u32) -> Result<(), Error> {
+        let ioctl_query = ioctl_query_supported_brightness(self);
+        match ioctl_query {
+            Ok(levels) => {
+                let new_value = levels.get_nearest(percentage);
+                ioctl_set_display_brightness(self, new_value).map_err(|e| e.into())
+            }
+            Err(e) if e == ERROR_NOT_FOUND => {
+                let new_value = ddcci_get_monitor_brightness(self)
+                    .map(|b| b.percentage_to_current(percentage))?;
+                ddcci_set_monitor_brightness(self, new_value).map_err(|e| e.into())
+            }
+            Err(e) => Err(SysError::SettingBrightnessFailed {
+                device_name: self.device_name.clone(),
+                function_name: "IOCTL_VIDEO_QUERY_SUPPORTED_BRIGHTNESS",
+                last_error: e,
+            }
+            .into()),
+        }
+    }
+}
+
+pub fn brightness_devices() -> impl Stream<Item = Result<Brightness, SysError>> {
+    unsafe extern "system" fn enum_monitors(
+        arg1: HMONITOR,
+        _arg2: HDC,
+        _arg3: *mut RECT,
+        arg4: LPARAM,
+    ) -> BOOL {
+        let monitors = &mut *(arg4.0 as *mut Vec<HMONITOR>);
+        monitors.push(arg1);
+        return true.into();
+    }
+    let mut hmonitors = Vec::<HMONITOR>::new();
+    let devices = unsafe {
+        if !EnumDisplayMonitors(
+            HDC::NULL,
+            std::ptr::null_mut() as *mut RECT,
+            Some(enum_monitors),
+            LPARAM(&mut hmonitors as *mut _ as isize),
+        )
+        .as_bool()
+        {
+            either::Left(iter::once(Err(SysError::ListingDevicesFailed(
+                "EnumDisplayMonitors",
+                GetLastError(),
+            ))))
+        } else {
+            either::Right(hmonitors.into_iter().flat_map(|hmonitor| {
+                // Get the Name of the HMONITOR
+                let mut info: MONITORINFOEXW = std::mem::MaybeUninit::zeroed().assume_init();
+                info.__AnonymousBase_winuser_L13558_C43.cbSize =
+                    std::mem::size_of::<MONITORINFOEXW>() as u32;
+                let info_ptr = &mut info as *mut _;
+                if !GetMonitorInfoW(hmonitor, info_ptr as *mut MONITORINFO).as_bool() {
+                    return vec![Err(SysError::ListingDevicesFailed(
+                        "GetMonitorInfoW",
```

Comment:
May you define different variants for different errors rather than having this string field indicating which function failed? This allows more precise error messages and the potential to add custom data if needed.